platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,852 @@
+valid path parameters. `/my-:index-item` is an invalid path parameter.
+
+`pathParams` will be an object whose keys map to the path parameters and whose
+values map to the path to the form data. In the above example, the `pathParams`
+will assign `:marriageIndex` in the URL a number based in the current item in
+the `marriages` array.
+
+##### Review page
+To keep Formulate simple, it will not build the review page automatically. It
+will be the responsibility of the engineers to build the review page. This
+decision is subject to change, but as of the time of this writing, I can't think
+of a way to automatically build the review page and meet all of the following
+requirements:
+- The chapters are divided correctly
+- The pages are divided correctly
+- The edit mode for each page is identical to the page in the form flow
+- The review mode for each page displays each field on the page in a data table
+  with the appropriate label and value
+- Each field in the review mode may be overridden when the label or value need
+  to be changed for a better user experience
+- Each page may or may not have a review mode
+
+Instead, Formulate will supply components to aid engineers in rapidly building
+review pages. These components will include:
+
+- `FormReview`
+- `ChapterReview`
+- `PageReview`
+- `FieldReview`
+
+**TODO:** Figure out how submission validation plays in here. How do we run all
+validation on all pages?
+
+#### Save-in-progress
+Pieces of the save-in-progress sub-module are:
+- [`SaveInProgress` component](#component-saveinprogress)
+  - Adds `SiPIndicator` to each page
+  - Provides functions to save and load the form
+- [`SiPIndicator` component](#component-sipindicator)
+  - Allows users to manually save the form
+  - Shows the form save state (saved, saving, error, etc.)
+- [Re-usable `FormSaved` component](#component-formsaved)
+  - To be used in a `<Route path="/form-saved">`
+- [Data migration functionality](#data-migrations)
+
+##### Component: `SaveInProgress`
+**Purpose:** This is the primary component for enabling the save in progress
+functionality for a form.
+
+**Usage requirements:**
+This component must be wrap the `Router` or `Formik` components so we can pass a
+`saveForm` prop to it to be hooked into `Formik`'s `handleChange` callback.
+
+**Responsibilities:**
+- Add the `SiPIndicator` to the bottom of each form page
+- Accept a list of paths to not display `SiPIndicator`
+- Provide a React context with the following
+  - `saveForm` function
+    - Save the form when `saveForm` is called
+  - `loadForm` function
+    - Call the supplied endpoint with the supplied arguments
+    - Set the form data to the form data in the result
+    - Perform any data migrations applicable
+    - Navigate to the page supplied by the result
+- Accept a [`dataMigrations`](#data-migrations) prop to be used with `loadForm`
+
+##### Component: `SiPIndicator`
+**TODO:** Rename this thing. Ideas?
```

Comment:
`<SaveStatus>`?

Reply:
Yeah, I thought about it, but it also displays the `Save your progress` link. (Or whatever the text is.) So it's not _just_ status.

Probably still better than `SiPIndicator`, so if I don't get any better suggestions, I'll change it to that, but I'll hold out for a while for better suggestions. :sweat_smile: